### PR TITLE
Feature/caching

### PR DESCRIPTION
**Pull request by 0c370t**

### CODE REVIEW

File: src/sprocket-common/caching/Cache.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+// eslint-disable-next-line arrow-body-style
+export const Cache: () => MethodDecorator = ({ttl}: CacheOptions = {ttl: 500}) => {
+    return (target: Object, propertyKey: string | symbol, descriptor: PropertyDescriptor) => {
+        const originalFunction = descriptor.value;
+        const cacheKeys: number[] = Reflect.getMetadata(cacheKeyMetadatakey, target) ?? [];
+
+
+        descriptor.value = async (...args: any[]) => {
+            const cacheValues: string[] = cacheKeys.map(ck => args[ck]?.toString() ?? "");
```
Comment (zachstence) [inline src/sprocket-common/caching/Cache.decorator.ts:41]: If an argument serializes to `[object Object]` we might have key collision

**Comment by 0c370t** [comment]
example endpoint that uses caching here

```typescript
  @Get()
  @Cache({
    ttl: 100000,
    transformers: {
      someValue: (v) => v.q,
    },
  })
  someEndpoint(
    @Query()
    @CacheKey
    someValue: any,
  ): string {
    console.log(someValue);
    return `${Date.now()} | ${someValue.q} | ${someValue.t}`;
  }

```

**Comment by zachstence** [comment]
> example endpoint that uses caching here
> 
> ```ts
>   @Get()
>   @Cache({
>     ttl: 100000,
>     transformers: {
>       someValue: (v) => v.q,
>     },
>   })
>   someEndpoint(
>     @Query()
>     @CacheKey
>     someValue: any,
>   ): string {
>     console.log(someValue);
>     return `${Date.now()} | ${someValue.q} | ${someValue.t}`;
>   }
> ```

Is there no way to specify the transformer on the `CacheKey` decorator so its right there next to the parameter?

File: src/sprocket-common/caching/Cache.decorator.ts [hunk 2]
```diff
@@ -31,23 +36,58 @@ async function getRedisClient(): Promise<Redis> {
...
         descriptor.value = async (...args: any[]) => {
-            const cacheValues: string[] = cacheKeys.map(ck => args[ck]?.toString() ?? "");
+            const cacheValues: string[] = cacheKeys.map(keyIndex => {
+                // Lookup the name of the key in question
+                const keyName = paramNames[keyIndex].toString();
+                // Apply transformation function if it exists
+                // eslint-disable-next-line no-prototype-builtins,@typescript-eslint/no-unsafe-argument
+                if (keyName && transformers?.hasOwnProperty(keyName)) return transformers[keyName](args[keyIndex]);
+                // Otherwise, return direct to string or empty string.
+                return args[keyIndex]?.toString() ?? "";
```
[the remark this answers is not in the document]
Comment (0c370t) [reply src/sprocket-common/caching/Cache.decorator.ts:72]: I don't think it's that big of an issue; at most I would say log it

**Comment by 0c370t** [comment]
> Is there no way to specify the transformer on the CacheKey decorator so its right there next to the parameter?

There is not one that I am aware of; you can't serialize/deserialize a function and class metadata is required to be a string

File: src/sprocket-common/caching/Cache.decorator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zachstence) [reply src/sprocket-common/caching/Cache.decorator.ts:72]: Using an empty string for a parameter would ruin the caching though, we would have conflicts which could lead to incorrect data being returned.

**Pull request merged**